### PR TITLE
new DDPG algorithm

### PR DESCRIPTION
**Pull request by pjfernandecabo**
**adding new deep deterministic policy gradient algorithm (DDPG)**

**Objective**: the DDPG algorithm belongs to the family of algorithms that try to find the most optimal policy. It is designed with actor-critic architecture:
Actor - It proposes an action given a state.
Critic - It predicts if the action is good (positive value) or bad (negative value) given a state and an action.

Both the actor and the critic are implemented with deep neural networks, which allow us to work with discrete or continuous actions and with the camera sensor as input to multidimensional state space.

**Files Modification/creation:**

1. Main directory:
- _/rl_studio/config.yaml_: is modified to introduce all the new DDPG algorithm hyperparameters

- _/rl_studio/main_rlstudio.py_: new lines in the get_environmet() function to get the actions from the yaml file

2. agents directory:

- _/rl_studio/agents/__init__.py_: Modification of the TrainerFactory class to exec the new DDPG algorithm

- new file _/rl_studio/agent/algorithms_type.py_ file: each different agent should run different RL algorithms like qlearn, dqn, sarsa, ddpg...

- new file _/rl_studio/agents/f1/train_ddpg.py_:  the DDPG trainer

3. Envs directory:

- new file _/rl_studio/envs/f1/models/f1_env_ddpg.py_: manages the DDPG algorithm environment (used for camera sensor or any other sensor)

- _/rl_studio/envs/f1/env_type.py_: a new line is created for the DDPG algorithm

- _/rl_studio/envs/f1/models/__init__.py_: the new DDPG environment is added to the F1Env class

4. algorithms directory:
-new file _/rl_studio/algorithms/ddpg.py_: contains all the logic and classes of the DDPG algorithm

### CODE REVIEW

File: rl_studio/agents/__init__.py [hunk 1]
```diff
@@ -1,20 +1,28 @@
...
-            from rl_studio.agents.f1.train_qlearn import F1Trainer
-            return F1Trainer(config)
+
+            if algorithm == AlgorithmType.QLEARN.value:
+                from rl_studio.agents.f1.train_qlearn import F1Trainer
+                return F1Trainer(config)
+                
+            elif algorithm == AlgorithmType.DDPG.value:
+                from rl_studio.agents.f1.train_ddpg import F1TrainerDDPG
+                return F1TrainerDDPG(config)
```
Comment (igarag) [inline rl_studio/agents/__init__.py:20]: We should move this code to the Algorithm factory. You can move it to [this file](https://github.com/JdeRobot/RL-Studio/blob/main/rl_studio/algorithms/__init__.py). Thus, we have separate the "Algoritm" and the "Agents".

How do you see it?

File: rl_studio/agents/f1/train_ddpg.py [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+import gym
+import pandas as pd
+from algorithms.ddpg import ModifiedTensorBoard, OUActionNoise, Buffer, DDPGAgent
+from envs.gazebo_env import *
+
+import tensorflow as tf
+from tensorflow import keras
+from tensorflow.keras import Input, Model, layers
+from tensorflow.keras.models import Sequential
+from tensorflow.keras.layers import Dense, Dropout, Conv2D, MaxPooling2D, Activation, Flatten, BatchNormalization, Lambda
```
Comment (igarag) [inline rl_studio/agents/f1/train_ddpg.py:23]: Same here. You are mixing the "agent" and the "algorithm". We must separate the code.

File: rl_studio/agents/f1/train_ddpg.py [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+    file_excel = f"{outdir_episode}/{current_time}_Circuit-{config['circuit']}_States-{config['state_space']}_Actions-{config['action_space']}_rewards-{config['reward_function']}.xlsx"
+    #writer = csv.writer(file)
+    #print(f"file:{file_csv}")
+
+    #for key, value in aggr_ep_rewards.items():
+    #    writer.writerow([key, value])
+    df = pd.DataFrame(aggr_ep_rewards)
+    df.to_csv(file_csv, mode='a', index = False, header=None)
+    #with pd.ExcelWriter(file_excel) as writer:
+    df.to_excel(file_excel)
```
Comment (igarag) [inline rl_studio/agents/f1/train_ddpg.py:52]: Remove comments that don't provide information

File: rl_studio/config.yaml [hunk 4]
```diff
@@ -2,18 +2,67 @@ settings:
...
+        # but remember image_resizing: it is easy getting out of image borders
+      sp5: 
+        0: [3, 5, 10, 15, 20]  
+      spn: 
+        0: [10] # Calculating center   
+    rewards:
+      reward_function: discrete # discrete, linear
+      discrete:
+        from_0_to_02: 10 #discrete ad hoc
+        from_02_to_04: 2 #discrete ad hoc
```
Comment (igarag) [inline rl_studio/config.yaml:51]: Move all the comments to a external document that contains the relevant information for the DDPG.

**Review by igarag — CHANGES_REQUESTED** [review]
Check the commens @pjfernandecabo. If you have any questions we can discuss it here :-)

File: rl_studio/agents/f1/train_ddpg.py (hunk 2 above)
Comment (pjfernandecabo) [reply rl_studio/agents/f1/train_ddpg.py:23]: I believe this file follows your current structure.
You have:
/agents/f1/train_qlearn.py
and
/algorithms/qlearn.py
So, when you want to train a F1 agent, the train_qlearn.py is calling the appropriate algorithm (learn).

Thus, now, 
/agents/f1/train_ddpg.py which is calling 
/algorithms/ddp.py

File: rl_studio/config.yaml (hunk 4 above)
Comment (pjfernandecabo) [reply rl_studio/config.yaml:51]: Okay, I'll remove the comments.
Comments could also be written in front of each section e.g., at the beginning of settings, agent, actions, environment and algorithm:

#####################
# SETTINGS
# in this section, define commons params
#
############################

###################
# AGENT
# F1 agent
# Camera_params defines...

and so on. I think it is pretty standard

**Comment by pjfernandecabo** [comment]
moving into new PR: f1_ddpg
